crates/subspace-networking/src/node_runner.rs

```diff
@@ -48,11 +51,32 @@ enum QueryResultSender {
     },
 }
 
+/// Perform arbitrary execution for ProviderRecord.
+pub trait ProviderRecordProcessor {
+    fn process_provider_record(&self, record: ProviderRecord);
+}
+
+impl<T: ProviderRecordProcessor> ProviderRecordProcessor for Option<T> {
+    fn process_provider_record(&self, record: ProviderRecord) {
+        if let Some(processor) = self {
+            processor.process_provider_record(record);
+        }
+    }
+}
```

Comment:
I would strongly prefer to not have such impls in our codebase. They seem convenient to write and end up confusing to read later.